### PR TITLE
Add node affinity

### LINKED ISSUE #511: Add nodeaffinity to prevent scheduling on virtual-kubelet nodes
**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]
Add nodeaffinity to prevent the CSI driver pods from being scheduled on [virtual-kubelet](https://github.com/virtual-kubelet/virtual-kubelet) nodes.

Using [type: virtual-kubelet](https://github.com/virtual-kubelet/virtual-kubelet/blob/master/cmd/virtual-kubelet/internal/commands/root/node.go#L43) label.

```yaml
affinity:
  nodeAffinity:
    requiredDuringSchedulingIgnoredDuringExecution:
      nodeSelectorTerms:
        - key: type
          operator: NotIn
          values:
          - virtual-kubelet
```

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]
We should approach this by adding an `affinity` field in the values.yaml and then setting this affinity by default in the values file. 

**Environment:**

- Secrets Store CSI Driver version: (use the image tag):
- Kubernetes version: (use `kubectl version`):

### PR DESCRIPTION
**Pull request by manedurphy**
**What this PR does / why we need it**:

This PR changes the behavior of scheduling for the CSI driver. The CSI driver should no long be scheduled on [virtual-kublet](https://github.com/virtual-kubelet/virtual-kubelet) nodes.

**Which issue(s) this PR fixes** *(optional, using `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when the PR gets merged)*:
Fixes #511

**Special notes for your reviewer**:
I have left a screen shot to show that a node with label `type=virtual-kubelet` was not receiving pods. This is my first contribution so please let me know if anything was done incorrectly or if I misunderstood anything. Thank you!

addNodeAffinityResult

### CODE REVIEW

File: manifest_staging/charts/secrets-store-csi-driver/templates/secrets-store-csi-driver-windows.yaml [hunk 1]
```diff
@@ -28,6 +28,8 @@ spec:
 {{- end }}
     spec:
       serviceAccountName: secrets-store-csi-driver
+      affinity:
+{{ toYaml .Values.affinity | indent 8 }}
```
Comment (aramase) [inline manifest_staging/charts/secrets-store-csi-driver/templates/secrets-store-csi-driver-windows.yaml:32]: We should have a separate affinity field for linux and windows. For example:

https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/master/charts/secrets-store-csi-driver/values.yaml#L120-L121

Comment (aramase) [inline manifest_staging/charts/secrets-store-csi-driver/templates/secrets-store-csi-driver-windows.yaml:32]:
```suggestion
{{ toYaml .Values.windows.affinity | indent 8 }}
```

File: manifest_staging/charts/secrets-store-csi-driver/templates/secrets-store-csi-driver.yaml [hunk 2]
```diff
@@ -28,6 +28,8 @@ spec:
 {{- end }}
     spec:
       serviceAccountName: secrets-store-csi-driver
+      affinity:
+{{ toYaml .Values.affinity | indent 8 }}
```
Comment (aramase) [inline manifest_staging/charts/secrets-store-csi-driver/templates/secrets-store-csi-driver.yaml:32]:
```suggestion
{{ toYaml .Values.linux.affinity | indent 8 }}
```

File: manifest_staging/charts/secrets-store-csi-driver/values.yaml [hunk 3]
```diff
@@ -152,3 +153,14 @@ rotationPollInterval:
 
 ## Filtered watch nodePublishSecretRef secrets
 filteredWatchSecret: false
+
+## Prevent the CSI driver from being scheduled on virtual-kublet nodes
+affinity:
```
Comment (aramase) [inline manifest_staging/charts/secrets-store-csi-driver/values.yaml:158]: We can nest this field inside linux and windows fields in the `values.yaml`

File: manifest_staging/charts/secrets-store-csi-driver/values.yaml [hunk 4]
```diff
@@ -152,3 +153,14 @@ rotationPollInterval:
...
+## Prevent the CSI driver from being scheduled on virtual-kublet nodes
+affinity: 
+  nodeAffinity:
+    requiredDuringSchedulingIgnoredDuringExecution:
+      nodeSelectorTerms:
+      - matchExpressions:
+        - key: type
+          operator: NotIn
+          values:
+          - virtual-kubelet
```
Comment (aramase) [inline manifest_staging/charts/secrets-store-csi-driver/values.yaml:166]: nit: new line

**Review by aramase — COMMENTED** [review]
@manedurphy The PR looks great! Just added a few comments for having the affinity fields in linux and windows yaml keys instead of having a global one. That'll allow users to set different affinity for linux and windows daemonsets.

File: manifest_staging/charts/secrets-store-csi-driver/values.yaml (hunk 3 above)
Comment (manedurphy) [reply manifest_staging/charts/secrets-store-csi-driver/values.yaml]: Hi @aramase,

Thank you for your feedback, I've added the changes here: https://github.com/kubernetes-sigs/secrets-store-csi-driver/pull/513/commits/b235b2222c3724a39308eb25c3873b652871504d

**Comment by aramase** [comment]
@manedurphy Looks like the rebase added already existing commits from master:
https://github.com/kubernetes-sigs/secrets-store-csi-driver/pull/513/commits/8941482d8dcb2f4b938590b7b36754ee0eb915da
https://github.com/kubernetes-sigs/secrets-store-csi-driver/pull/513/commits/deb5476f0e8e9ab055213521f61c50c9190470ae

Could you rebase to just include your commit and also squash it? 

This is a useful doc for contribution: https://www.kubernetes.dev/docs/guide/github-workflow/. Most of the guidelines also apply to all the repos in kubernetes/kubernetes-sigs. Please feel free to ping if you have any questions.

**Comment by manedurphy** [comment]
@aramase yes I will make this correction. I suspected this might need revision. I may reach out if I encounter any issues, but will have a go at it myself first.

File: manifest_staging/charts/secrets-store-csi-driver/README.md [hunk 5]
```diff
@@ -26,6 +26,7 @@ The following table lists the configurable parameters of the csi-secrets-store-p
 | `linux.image.repository`                | Linux image repository                                                                                                            | `k8s.gcr.io/csi-secrets-store/driver`               |
 | `linux.image.pullPolicy`                | Linux image pull policy                                                                                                           | `Always`                                            |
 | `linux.image.tag`                       | Linux image tag                                                                                                                   | `v0.0.21`                                           |
+| `linux.affinity.nodeAffinity.nodeSelectorTerms`                         | Linux affinity                                                                                     | `key: type; operator: NotIn; values: [virtual-kubelet]`|
```
Comment (aramase) [inline manifest_staging/charts/secrets-store-csi-driver/README.md:29]:
```suggestion
| `linux.affinity`                         | Linux affinity                                                                                     | `key: type; operator: NotIn; values: [virtual-kubelet]`| 
```

File: manifest_staging/charts/secrets-store-csi-driver/README.md [hunk 6]
```diff
@@ -51,6 +52,7 @@ The following table lists the configurable parameters of the csi-secrets-store-p
 | `windows.image.repository`              | Windows image repository                                                                                                          | `k8s.gcr.io/csi-secrets-store/driver`               |
 | `windows.image.pullPolicy`              | Windows image pull policy                                                                                                         | `IfNotPresent`                                      |
 | `windows.image.tag`                     | Windows image tag                                                                                                                 | `v0.0.21`                                           |
+| `windows.affinity.nodeAffinity.nodeSelectorTerms`                         | Windows affinity                                                                                 | `key: type; operator: NotIn; values: [virtual-kubelet]`|
```
Comment (aramase) [inline manifest_staging/charts/secrets-store-csi-driver/README.md:55]:
```suggestion
| `windows.affinity`                         | Windows affinity                                                                                 | `key: type; operator: NotIn; values: [virtual-kubelet]`|
```

Comment (manedurphy) [reply manifest_staging/charts/secrets-store-csi-driver/README.md]: changes made

**Review by aramase — COMMENTED** [review]
/ok-to-test

**Review by aramase — COMMENTED** [review]
Thank you for the PR @manedurphy!

/approve

**Pull request merged**